### PR TITLE
[MANA-61] Switch from upper half when processing signals.

### PR DESCRIPTION
**Pull request by karya0**

### CODE REVIEW

**Comment by karya0** [comment]
@JainTwinkle : Added the #ifdef in the code now. Please take a look.

File: mpi-proxy-split/mpi_plugin.cpp [hunk 1]
```diff
@@ -149,6 +154,93 @@ dmtcp_skip_memory_region_ckpting(const ProcMapsArea *area)
...
+  if (signum == dmtcp_get_ckpt_signal() ||
+      signum >= MaxSignals) {
+    return NEXT_FNC(sigaction)(signum, act, oldact);
+  }
+
+  if (act != NULL) {
+    struct sigaction newAct = *act;
+    if (newAct.sa_flags & SA_SIGINFO) {
+      newAct.sa_sigaction = mana_signal_sa_sigaction_wrapper;
+      ret = NEXT_FNC(sigaction)(signum, &newAct, oldact);
```
Comment (JainTwinkle) [inline mpi-proxy-split/mpi_plugin.cpp:216]: oldact -> oldAct for consistency.

File: mpi-proxy-split/mpi_plugin.cpp [hunk 2]
```diff
@@ -461,17 +553,40 @@ mpi_plugin_event_hook(DmtcpEvent_t event, DmtcpEventData_t *data)
   switch (event) {
     case DMTCP_EVENT_INIT: {
       JTRACE("*** DMTCP_EVENT_INIT");
+      JASSERT(dmtcp_get_real_tid != NULL);
+      initialize_signal_handlers();
       initialize_segv_handler();
       JASSERT(!splitProcess()).Text("Failed to create, initialize lower haf");
```
Comment (JainTwinkle) [inline mpi-proxy-split/mpi_plugin.cpp:559]: Let's fix a typo here!
haf -> half

File: mpi-proxy-split/mpi_plugin.cpp [hunk 3]
```diff
@@ -461,17 +553,40 @@ mpi_plugin_event_hook(DmtcpEvent_t event, DmtcpEventData_t *data)
...
     }
     case DMTCP_EVENT_EXIT: {
       JTRACE("*** DMTCP_EVENT_EXIT");
       seq_num_destroy();
       break;
     }
+
+    case DMTCP_EVENT_PTHREAD_START: {
+      // Do we need a mutex here?
+      DmtcpMutexLock(&g_upper_half_fsbase_lock);
```
Comment (JainTwinkle) [inline mpi-proxy-split/mpi_plugin.cpp:575]: tid would be unique for each thread. So, I guess we can skip the mutex here.

File: mpi-proxy-split/mpi_plugin.cpp [hunk 4]
```diff
@@ -461,17 +553,40 @@ mpi_plugin_event_hook(DmtcpEvent_t event, DmtcpEventData_t *data)
...
+      g_upper_half_fsbase[dmtcp_get_real_tid()] = getFS();
+      DmtcpMutexUnlock(&g_upper_half_fsbase_lock);
+      break;
+    }
+
+    case DMTCP_EVENT_PTHREAD_RETURN:
+    case DMTCP_EVENT_PTHREAD_EXIT: {
+      // Do we need a mutex here?
```
Comment (JainTwinkle) [inline mpi-proxy-split/mpi_plugin.cpp:583]: Same as above!

File: mpi-proxy-split/mpi_plugin.cpp (hunk 3 above)
Comment (karya0) [reply mpi-proxy-split/mpi_plugin.cpp:575]: STL `unordered_map` is not thread-safe so that's why we need a mutex.

Comment (JainTwinkle) [reply mpi-proxy-split/mpi_plugin.cpp:575]: I believe different threads can access and modify different elements concurrently. In our case, unique TIDs should be enough to ensure that no two threads will access the same key.

Comment (karya0) [reply mpi-proxy-split/mpi_plugin.cpp:575]: They can access it but if one thread tries to modify unordered_map (insert, erase), that'll cause a race condition and the results are unexpected. In multi-threaded application, STL containers need to be protected with mutex.

Comment (JainTwinkle) [reply mpi-proxy-split/mpi_plugin.cpp:575]: Got it.

**Review by yuelimv — APPROVED** [review]

**Pull request merged**